File: volterra-braid/oracle/bench_braid.py
"""Performance benchmark: volterra braid pipeline (Rust/PyO3) vs the CGPO Python.

Times the full braid-analysis pipeline -- defect detection from Q-tensor grids,
tracking, and braid-word extraction -- on identical input, comparing:

  - volterra (Rust, via the `volterra` PyO3 extension), and
  - braid_tracker_v2.py (a faithful pure-Python transcription of the CGPO
    braid_tracker.py algorithm; its per-cell ss loop + flood fill + tracking are
    the same work the published script does, minus the plotting/IO).

Detection dominates (an Lx*Ly per-cell Jacobian + connected-component flood fill
per frame), so this is essentially a native-vs-Python throughput comparison of
the defect-tracking analytics.

Run (needs `maturin develop --release` first):
    .venv/bin/python volterra-braid/oracle/bench_braid.py
    .venv/bin/python volterra-braid/oracle/bench_braid.py 150 128   # frames grid
"""

from __future__ import annotations

import sys
import time

import braid_tracker_v2 as v2
import compare_cgpo as cc

try:
    import volterra
except ImportError:
    sys.exit("ERROR: import volterra failed; run `maturin develop --release` first.")


def build_grids(n_frames, lx):
    """Render `n_frames` Q-grids of a 3-defect golden orbit at lx*lx resolution."""
    cc.LX = cc.LY = lx
    cc.RADIUS = lx // 2 - 1
    # Scale the orbit geometry to the grid so defects stay well inside.
    cc.GY_SPACING = {3: lx * 0.24, 4: lx * 0.2}
    pts = cc.disk_orbit(3, [-2, 1], fpg=10, periods=max(1, n_frames // 20))
    pts = pts[:n_frames]
    grids = [cc.render_q(p) for p in pts]  # list of (qxx_flat, qxy_flat)
    mask = cc.circular_mask()
    return grids, mask


def time_volterra(grids, mask, lx):
    frames = []
    t0 = time.perf_counter()
    for (qxx, qxy) in grids:
        defs = volterra.braid_detect_defects(qxx.tolist(), qxy.tolist(), lx, lx, cc.THRESHOLD, mask)
        frames.append([(y, x, c) for (x, y, c) in defs])  # (gy, gx) ordering
    t_det = time.perf_counter() - t0
    t1 = time.perf_counter()
    _, codes = volterra.braid_word_from_frames(frames)
    t_word = time.perf_counter() - t1
    return t_det, t_word, codes


def time_python(grids, mask, lx):
    frames = []
    t0 = time.perf_counter()
    for (qxx, qxy) in grids:
        defs = v2.detect_defects(qxx.tolist(), qxy.tolist(), lx, lx, cc.THRESHOLD, mask)
        frames.append([(y, x, c) for (x, y, c) in defs])
    t_det = time.perf_counter() - t0
    t1 = time.perf_counter()
    _, codes = v2.braidword_from_frames(frames)
    t_word = time.perf_counter() - t1
    return t_det, t_word, codes


def main():
    n_frames = int(sys.argv[1]) if len(sys.argv) > 1 else 120
    lx = int(sys.argv[2]) if len(sys.argv) > 2 else 100
    print(f"Braid pipeline benchmark: {n_frames} frames at {lx}x{lx} ({lx*lx} sites/frame)\n")

    grids, mask = build_grids(n_frames, lx)

    # Warm-up (JIT-free, but warms caches / first-call allocation).
    time_volterra(grids[:5], mask, lx)
    time_python(grids[:5], mask, lx)

    v_det, v_word, v_codes = time_volterra(grids, mask, lx)
    p_det, p_word, p_codes = time_python(grids, mask, lx)

    sites = n_frames * lx * lx
    print(f"{'stage':<22}{'volterra (Rust)':>20}{'braid_tracker_v2 (Py)':>24}{'speedup':>10}")
    print("-" * 76)
    for label, vt, pt in (("detection (total)", v_det, p_det), ("track+word (total)", v_word, p_word)):
        print(f"{label:<22}{vt*1e3:>17.2f} ms{pt*1e3:>21.2f} ms{pt/max(vt,1e-12):>9.1f}x")
    v_tot, p_tot = v_det + v_word, p_det + p_word
    print(f"{'TOTAL':<22}{v_tot*1e3:>17.2f} ms{p_tot*1e3:>21.2f} ms{p_tot/max(v_tot,1e-12):>9.1f}x")
    print()
    print(f"  volterra: {v_det/n_frames*1e6:8.2f} us/frame detection, {v_det/sites*1e9:6.2f} ns/site")
    print(f"  python:   {p_det/n_frames*1e6:8.2f} us/frame detection, {p_det/sites*1e9:6.2f} ns/site")
    print(f"\n  braid words match: {v_codes == p_codes}  (volterra {v_codes[:6]}...)")


if __name__ == "__main__":
    main()

File: volterra-braid/examples/bench_braid.rs
//! Native throughput benchmark for the braid-analysis pipeline (no Python/PyO3).
//!
//! Renders Q-tensor grids of a moving 3-defect configuration, then times defect
//! detection, tracking, and braid-word extraction in pure Rust.
//!
//! Run: `cargo run --release --example bench_braid -p volterra-braid`

use std::time::Instant;
use volterra_braid::{detect_defects, extract_braidword, track};

const LX: usize = 100;
const THRESHOLD: f64 = 0.1;

/// Superpose +1/2 winding fields at `defects` -> (qxx, qxy), row-major x*LX+y.
fn render(defects: &[(f64, f64)]) -> (Vec<f64>, Vec<f64>) {
    let mut qxx = vec![0.0; LX * LX];
    let mut qxy = vec![0.0; LX * LX];
    for x in 0..LX {
        for y in 0..LX {
            let mut ang = 0.0;
            for &(px, py) in defects {
                ang += (y as f64 - py).atan2(x as f64 - px);
            }
            qxx[x * LX + y] = ang.cos();
            qxy[x * LX + y] = ang.sin();
        }
    }
    (qxx, qxy)
}

/// One-cell border ring zeroed (true = active interior), matching the harness.
fn interior_mask() -> Vec<bool> {
    let r = (LX / 2 - 1) as f64;
    let c = (LX / 2 - 1) as f64;
    let mut m = vec![false; LX * LX];
    for x in 0..LX {
        for y in 0..LX {
            let inside = (x as f64 - c).powi(2) + (y as f64 - c).powi(2) <= r * r;
            let border = x == 0 || y == 0 || x == LX - 1 || y == LX - 1;
            m[x * LX + y] = inside && !border;
        }
    }
    m
}

/// `n_defects` sliding along grid-y (spaced), each with a small grid-x wobble.
/// Exact braid is irrelevant for timing; this just exercises detection +
/// tracking + extraction with the right number of defects.
fn make_grids(n_defects: usize, n_frames: usize) -> Vec<(Vec<f64>, Vec<f64>)> {
    let cx = (LX / 2 - 1) as f64;
    let span = (LX as f64) * 0.44;
    let mut grids = Vec::with_capacity(n_frames);
    for f in 0..n_frames {
        let t = f as f64 / n_frames as f64;
        let defects: Vec<(f64, f64)> = (0..n_defects)
            .map(|k| {
                let gy = cx + (k as f64 - (n_defects as f64 - 1.0) / 2.0) * span / n_defects as f64;
                let phase = (k as f64) * 1.7 + t * std::f64::consts::TAU;
                (cx + 6.0 * phase.sin(), gy)
            })
            .collect();
        grids.push(render(&defects));
    }
    grids
}

fn bench(label: &str, n_defects: usize, n_frames: usize, mask: &[bool]) {
    let grids = make_grids(n_defects, n_frames);

    for (qxx, qxy) in grids.iter().take(5) {
        let _ = detect_defects(qxx, qxy, LX, LX, THRESHOLD, mask);
    }

    let t0 = Instant::now();
    let mut frames = Vec::with_capacity(n_frames);
    for (qxx, qxy) in &grids {
        frames.push(detect_defects(qxx, qxy, LX, LX, THRESHOLD, mask));
    }
    let t_det = t0.elapsed();

    let t1 = Instant::now();
    let word = extract_braidword(&track(&frames));
    let t_word = t1.elapsed();

    let sites = (n_frames * LX * LX) as f64;
    let det_s = t_det.as_secs_f64();
    let counts: std::collections::BTreeSet<usize> = frames.iter().map(|f| f.len()).collect();
    println!("\n=== {label} ({n_defects} defects) ===");
    println!(
        "  detection:   {:8.2} ms total, {:7.2} us/frame, {:6.2} ns/site  (defects/frame {:?})",
        det_s * 1e3,
        det_s / n_frames as f64 * 1e6,
        det_s / sites * 1e9,
        counts,
    );
    println!(
        "  track+word:  {:8.3} ms total ({} generators over {} frames)",
        t_word.as_secs_f64() * 1e3,
        word.gens.len(),
        n_frames,
    );
}

fn main() {
    let n_frames = 120usize;
    let mask = interior_mask();
    println!("Native braid pipeline: {n_frames} frames at {LX}x{LX}");
    bench("golden", 3, n_frames, &mask);
    bench("silver", 4, n_frames, &mask);
}
